[TACHYON-646] Use createNewClassInstance to improve Allocator factory

sandihouse opened the pull request:
https://tachyon.atlassian.net/browse/TACHYON-646
1. use createNewClassInstance to implement new Allocator$Factory
2. remove unreferenced AllocatorFactory class and AllocatorType class

AmplabJenkins commented:
Can one of the admins verify this patch?

gpang commented:
ok to test

apc999 commented on servers/src/test/java/tachyon/worker/block/allocator/AllocatorFactoryTest.java:
```diff
@@ -50,19 +54,29 @@ public void before() throws IOException {
-  public void createDefaultAllocatorTypeTest() {
-    Allocator allocator = AllocatorFactory.create(AllocatorType.DEFAULT, mManagerView);
+  public void createDefaultAllocatorTest() {
+    TachyonConf conf = new TachyonConf();
```
Do we need a new `TachyonConf` instance here? Maybe just use `mTachyonConf` since it is initialized everytime in `before`

gpang replied:
I think it needs to be a new instance here, to test the default allocator.

Maybe leave a short comment about it?

apc999 commented on servers/src/main/java/tachyon/worker/block/allocator/Allocator.java:
```diff
@@ -25,11 +31,32 @@
+
+  class Factory {
+    /**
+     * Create a new {@link Allocator} instance
```
Could you javadoc here that, the returned `Allocator` is `MaxFreeAllocator` by default.

apc999 commented:
Mostly look good to me with two minor comments left. Thanks for the patch!

gpang commented on servers/src/test/java/tachyon/worker/block/allocator/AllocatorFactoryTest.java:
```diff
@@ -25,24 +25,28 @@
- * and test if it generate the correct Allocator instance.
+/**
+ * Test {@link Allocator.Factory} by passing different allocate strategy class
+ * name with tachyon conf and test if it generate the correct Allocator instance.
```
name -> names
generate -> generates

gpang commented on servers/src/test/java/tachyon/worker/block/allocator/AllocatorContractTest.java:
```diff
@@ -16,20 +16,58 @@
+        }
+      }
+    } catch (Exception e) {
+      Assert.assertTrue("Failed to find implementation of allocate strategy", false);
```
Assert.fail() is more appropriate.

apc999 commented:
@gpang, please review this PR

haoyuan commented:
We can’t automatically merge this pull request.

sandihouse commented:
Sorry, i have just rebased this PR on top of latest master and fixed the review commits after this weekend breaks, it should be available for automatically merge now

gpang commented:
Could you update the configuration docs to include the new changes from this PR? Thanks.

gpang commented on common/src/main/java/tachyon/Constants.java:
```diff
@@ -172,7 +172,7 @@
       "tachyon.worker.network.netty.shutdown.timeout";
   public static final String WORKER_EVICT_STRATEGY_TYPE = "tachyon.worker.evict.strategy";
-  public static final String WORKER_ALLOCATE_STRATEGY_TYPE = "tachyon.worker.allocate.strategy";
+  public static final String WORKER_ALLOCATE_STRATEGY = "tachyon.worker.allocate.strategy.class";
```
Maybe "WORKER_ALLOCATE_STRATEGY_CLASS" would be more appropriate?

gpang commented:
I just have a minor comment, but overall looks good. Thanks for the PR!

sandihouse commented:
@gpang  It's OK, thanks for your review comments, I have push a fix about that.

gpang commented:
LGTM

apc999 commented:
+1 LGTM
